File: solutions/typescript/rational-numbers/2/rational-numbers.ts
export class Rational {
  numerator: number
  denominator: number

  constructor(numerator: number, denominator: number) {
    this.numerator = numerator
    this.denominator = denominator
  }

  formatReal() {
      if (this.numerator === 0) {
          this.denominator = 1
      }
      if (this.denominator < 0) {
          this.numerator = -this.numerator
          this.denominator = -this.denominator
      }
      return this
  }

  add(other: Rational) {
    const numerator = this.numerator * other.denominator + other.numerator * this.denominator
    const denominator = this.denominator * other.denominator
    return new Rational(numerator, denominator).formatReal()
  }

  sub(other: Rational) {
    const numerator = this.numerator * other.denominator - other.numerator * this.denominator
    const denominator = this.denominator * other.denominator
    return new Rational(numerator, denominator).formatReal()
  }

  mul(other: Rational) {
      const numerator = this.numerator * other.numerator
      const denominator = this.denominator * other.denominator
      return new Rational(numerator, denominator).formatReal().reduce()
  }

  div(other: Rational) {
      const numerator = this.numerator * other.denominator
      const denominator = this.denominator * other.numerator
      return new Rational(numerator, denominator).reduce()
  }

  abs() {
    const numerator = Math.abs(this.numerator)
    const denominator = Math.abs(this.denominator)
    return new Rational(numerator, denominator).formatReal().reduce()
  }

  exprational(exp: number) {
    if (exp === 0) {
        return new Rational(1, 1)
    }
    if (exp < 0) {
        const numerator = Math.pow(this.denominator, Math.abs(exp))
        const denominator = Math.pow(this.numerator, Math.abs(exp))
        return new Rational(numerator, denominator).formatReal()
    }
    const numerator = Math.pow(this.numerator, exp)
    const denominator = Math.pow(this.denominator, exp)
    return new Rational(numerator, denominator).formatReal()
  }

  expreal(exp: number) {
      // raise the real number to the power of the rational number
      const numerator = Math.pow(exp, this.numerator)
      return Math.pow(numerator, 1 / this.denominator)
  }

  reduce() {
      const gcd = (a: number, b: number): number => b === 0 ? a: gcd(b, a % b)
      const divisor = gcd(this.numerator, this.denominator)
      const numerator = this.numerator / divisor
      const denominator = this.denominator / divisor
      return new Rational(numerator, denominator).formatReal()
  }
}

File: solutions/c/rational-numbers/2/rational_numbers.c
#include "rational_numbers.h"
#include <math.h>
#include <stdlib.h>

int _gcd_euclidean(int a, int b);

rational_t add(rational_t r1, rational_t r2) {
    rational_t  ret;

    ret.numerator = r1.numerator * r2.denominator + r2.numerator * r1.denominator;
    ret.denominator = r1.denominator * r2.denominator;

    return reduce(ret);
}

rational_t subtract(rational_t r1, rational_t r2) {
    rational_t ret;

    ret.numerator = r1.numerator * r2.denominator - r2.numerator * r1.denominator;
    ret.denominator = r1.denominator * r2.denominator;

    return reduce(ret);
}

rational_t multiply(rational_t r1, rational_t r2) {
    rational_t ret;

    ret.numerator = r1.numerator * r2.numerator;
    ret.denominator = r1.denominator * r2.denominator;

    return reduce(ret);
}

rational_t divide(rational_t r1, rational_t r2) {
    rational_t ret;

    ret.numerator = r1.numerator * r2.denominator;
    ret.denominator = r2.numerator * r1.denominator;

    return reduce(ret);
}

rational_t absolute(rational_t r) {
    rational_t ret = {abs(r.numerator), abs(r.denominator)};
    return reduce(ret);
}

rational_t exp_rational(rational_t r, uint16_t n) {
    rational_t ret = {pow(r.numerator, n), pow(r.denominator, n)};
    return reduce(ret);
}

float exp_real(float x, rational_t r) {
    return pow(pow(x, r.numerator), 1.0 / r.denominator);
}

int _gcd_euclidean(int a, int b) {
    int larger = a >= b ? a : b;
    int smaller = a <= b ? a : b;

    return smaller == 0 ? larger : _gcd_euclidean(smaller, larger % smaller );
}

rational_t reduce(rational_t r) {
    int gcd = _gcd_euclidean(abs(r.numerator), abs(r.denominator));

    rational_t ret = {r.numerator / gcd, r.denominator / gcd};

    if (ret.denominator < 0) {
        ret.numerator *= -1;
        ret.denominator *= -1;
    }

    return ret;
}